hail/src/main/scala/is/hail/expr/types/physical/PArray.scala

```diff
@@ -39,4 +39,42 @@ final case class PArray(elementType: PType, override val required: Boolean = fal
     assert(this isOfType other)
     CodeOrdering.iterableOrdering(this, other.asInstanceOf[PArray], mb)
   }
+
+  def checkedConvertFrom(mb: EmitMethodBuilder, r: Code[Region], value: Code[_], otherPT: PType, msg: String): Code[Long] = {
+    val otherPTA = otherPT.asInstanceOf[PArray]
+    assert(otherPTA.elementType.isPrimitive)
+    val oldOffset = coerce[Long](value)
+    val len = otherPTA.loadLength(oldOffset)
+    if (otherPTA.elementType.required == elementType.required) {
+      coerce[Long](value)
+    }
+    else {
+      if (otherPTA.elementType.required) {
+        // convert from required to non-required
+        val newOffset = mb.newField[Long]
+        Code(
+          newOffset := allocate(r, len),
+          stagedInitialize(newOffset, len),
+          Region.copyFrom(otherPTA.elementsOffset(len), newOffset, len.toL * elementByteSize),
+          newOffset
+        )
+      } else {
+        //  convert from non-required to required
+        val newOffset = mb.newField[Long]
+        val i = mb.newField[Int]
+        Code(
+          newOffset := allocate(r, len),
+          stagedInitialize(newOffset, len),
+          i := 0,
+          Code.whileLoop(i < len,
+            otherPTA.isElementMissing(oldOffset, i).orEmpty(Code._fatal(s"${msg}: convertFrom $otherPT failed: element missing.")),
+            i := i + 1
+          ),
+          Region.storeInt(newOffset.load(), len),
```

Comment:
This is handled by the `stagedInitialize`

Reply:
Who calls `stagedInitialize`?

Reply:
Oh I see what you mean